pymc_marketing/mmm/utility.py

```diff
@@ -0,0 +1,546 @@
+#   Copyright 2024 The PyMC Labs Developers
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+#   Copyright 2024 The PyMC Labs Developers
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+"""
+Utility functions for Bayesian optimization.
+
+Key Concepts:
+-------------
+
+- **Samples**:
+    A PyTensor tensor variable (`pt.TensorVariable`) representing samples drawn from the posterior
+    distributions of the model outputs. These samples capture the uncertainty in the model predictions
+    and are essential for computing expected utilities and risk measures in Bayesian optimization.
+
+- **Budgets**:
+    A PyTensor tensor variable representing a set of monetary budgets allocated to different assets,
+    investments, or channels. Each element corresponds to the budget for a specific option in the
+    optimization process.
+"""
+
+from collections.abc import Callable
+
+import pytensor.tensor as pt
+
+UtilityFunction = Callable[[pt.TensorVariable, pt.TensorVariable], float]
+
+
+def _compute_quantile(x: pt.TensorVariable, q: float) -> pt.TensorVariable:
+    """
+    Compute the quantile of a PyTensor tensor variable.
+
+    Parameters
+    ----------
+    x : pt.TensorVariable
+        A 1D PyTensor tensor variable containing samples.
+    q : float
+        The quantile to compute, between 0 and 1.
+
+    Returns
+    -------
+    pt.TensorVariable
+        The quantile value.
+    """
+    sorted_x = pt.sort(x)
+    n = x.shape[0]
+    idx = q * (n - 1)
+    idx_floor = pt.floor(idx).astype("int64")
+    idx_ceil = pt.ceil(idx).astype("int64")
+    weight = idx - idx_floor
+    return (1 - weight) * sorted_x[idx_floor] + weight * sorted_x[idx_ceil]
+
+
+def average_response(
+    samples: pt.TensorVariable, budgets: pt.TensorVariable
+) -> pt.TensorVariable:
+    """Compute the average response of the posterior predictive distribution."""
+    return pt.mean(samples)
```

Comment:
Do we really need this function? Note we are not using the `budget` variable at all, right?

Reply:
We need the same signature in all functions in order to make compatible an scalable. I had an implementation that was more directed per function but was taking a few flexibility from users, with @wd60622 we arrive to this conclusion with makes everything more flexible, the only inconvenience is we need `samples` and `budgets` across all, even when we don't use them.